### PR TITLE
Response Records

### LINKED ISSUE #3: Response Records
Use Hirefy document. TASK 1. Without CSV import API(another task).
- add all needed models and relations
- implement business logic for scoring

### PR DESCRIPTION
**Pull request by oskvortsov-bmt**
Closes #3 adding endpoint for saving data and another model

### CODE REVIEW

File: api/models/CandidateResponse.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+module.exports = {
+  schema: true,
+  attributes: {
+    id: {
+      type: "integer",
+      primaryKey: true
```
Comment (zaikin-andrew) [inline api/models/CandidateResponse.js]: autoIncrement: true

File: api/models/Job.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+module.exports = {
+  schema: true,
+  attributes: {
+    id: {
+      type: "integer",
+      primaryKey: true
```
Comment (zaikin-andrew) [inline api/models/Job.js]: autoIncrement: true

File: api/models/Options.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+module.exports = {
+  schema: true,
+  attributes: {
+    id: {
+      type: "integer",
+      primaryKey: true
```
Comment (zaikin-andrew) [inline api/models/Options.js]: autoIncrement: true

File: api/models/Options.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    },
+    id_question: {
+      model: 'Question',
```
Comment (zaikin-andrew) [inline api/models/Options.js]: use lower case for models' names

File: api/models/Options.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+module.exports = {
```
Comment (zaikin-andrew) [inline api/models/Options.js]: Any model is not plural.

File: api/models/CandidateResponse.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    },
+    timestamp: {
+      type: "string"
+    },
+    id_survey: {
+      model: "Survey",
+      via: "id"
+    },
+    id_job: {
```
Comment (zaikin-andrew) [inline api/models/CandidateResponse.js]: One-to-many. Just model here.

File: api/models/CandidateResponse.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  attributes: {
+    id: {
+      type: "integer",
+      primaryKey: true
+    },
+    timestamp: {
+      type: "string"
+    },
+    id_survey: {
+      model: "Survey",
```
Comment (zaikin-andrew) [inline api/models/CandidateResponse.js]: One-to-many. Just model here.

File: api/models/Options.js (hunk 5 above)
Comment (oskvortsov-bmt) [reply api/models/Options.js]: I don`t use name 'Option' because this name exists in framework

File: api/models/CandidateResponse.js [hunk 8]
```diff
@@ -3,18 +3,18 @@ module.exports = {
...
-      primaryKey: true
+      primaryKey: true,
+      autoIncrement: true
     },
     timestamp: {
       type: "string"
     },
-    id_survey: {
+    survey: {
       model: "Survey",
```
Comment (zaikin-andrew) [inline api/models/CandidateResponse.js]: lower case

File: api/models/CandidateResponse.js [hunk 9]
```diff
@@ -3,18 +3,18 @@ module.exports = {
...
-      via: "id"
     },
-    id_job: {
-      model: "Job",
-      via: "id"
+    responses: {
+      collection: "response",
+      via: 'candidate_response'
```
Comment (zaikin-andrew) [inline api/models/CandidateResponse.js]: camelCase

File: api/models/CandidateResponse.js [hunk 10]
```diff
@@ -3,18 +3,18 @@ module.exports = {
...
     }
   },
   tableName: "candidate_responses"
```
Comment (zaikin-andrew) [inline api/models/CandidateResponse.js]: camelCase

File: api/models/Job.js [hunk 11]
```diff
@@ -3,18 +3,19 @@ module.exports = {
   attributes: {
     id: {
       type: "integer",
-      primaryKey: true
+      primaryKey: true,
+      autoIncrement: true
     },
     job_name: {
```
Comment (zaikin-andrew) [inline api/models/Job.js]: camelCase

File: api/models/Job.js [hunk 12]
```diff
@@ -3,18 +3,19 @@ module.exports = {
...
       type: "string"
     },
-    id_survey: {
-      model: "Survey",
-      via: "id"
-    },
     id_customer: {
```
Comment (zaikin-andrew) [inline api/models/Job.js]: camelCase

File: api/models/Job.js [hunk 13]
```diff
@@ -3,18 +3,19 @@ module.exports = {
...
       type: "integer"
-    }
+    },
+    surveys: {
+      collection: "survey",
+      via: "job"
+    },
   },
   tableName: "jobs"
```
Comment (zaikin-andrew) [inline api/models/Job.js]: singular

File: api/models/Options.js [hunk 14]
```diff
@@ -3,17 +3,17 @@ module.exports = {
...
       type: "string"
     },
     score: {
       type: 'integer'
+    },
+    question: {
+      model: 'question'
     }
   },
   tableName: "options"
```
Comment (zaikin-andrew) [inline api/models/Options.js]: singular

File: api/models/Question.js [hunk 15]
```diff
@@ -3,11 +3,8 @@ module.exports = {
...
       type: "integer",
-      primaryKey: true
-    },
-    id_survey: {
-      model: "Survey",
-      via: "id"
+      primaryKey: true,
+      autoIncrement: true
     },
     id_group: {
```
Comment (zaikin-andrew) [inline api/models/Question.js]: camelCase

File: api/models/Question.js [hunk 16]
```diff
@@ -18,8 +15,15 @@ module.exports = {
...
+    },
     text: {
       type: "text"
+    },
+    options: {
+      collection: "options",
+      via: "question"
     }
   },
   tableName: "questions"
```
Comment (zaikin-andrew) [inline api/models/Question.js]: singular

File: api/models/Response.js [hunk 17]
```diff
@@ -3,22 +3,20 @@ module.exports = {
...
-    id_candidate_response: {
-      model: "CandidateResponse",
-      via: "id"
+    text: {
+      type: "string"
     },
-    id_question: {
-      model: 'Question',
-      via: "id"
+    candidate_response: {
```
Comment (zaikin-andrew) [inline api/models/Response.js]: camelCase

File: api/models/Response.js [hunk 18]
```diff
@@ -3,22 +3,20 @@ module.exports = {
...
+      model: "CandidateResponse"
     },
-    id_option: {
-      model: "Options",
-      via: "id"
+    question: {
+      model: 'Question'
```
Comment (zaikin-andrew) [inline api/models/Response.js]: lower case

File: api/models/Response.js [hunk 19]
```diff
@@ -3,22 +3,20 @@ module.exports = {
...
-      model: "CandidateResponse",
-      via: "id"
+    text: {
+      type: "string"
     },
-    id_question: {
-      model: 'Question',
-      via: "id"
+    candidate_response: {
+      model: "CandidateResponse"
```
Comment (zaikin-andrew) [inline api/models/Response.js]: lower case

File: api/models/Response.js [hunk 20]
```diff
@@ -3,22 +3,20 @@ module.exports = {
...
-    id_option: {
-      model: "Options",
-      via: "id"
+    question: {
+      model: 'Question'
     },
-    text: {
-      type: "string"
+    option: {
+      model: "Options"
```
Comment (zaikin-andrew) [inline api/models/Response.js]: lower case

File: api/models/Response.js [hunk 21]
```diff
@@ -3,22 +3,20 @@ module.exports = {
...
     }
   },
   tableName: "responses"
```
Comment (zaikin-andrew) [inline api/models/Response.js]: singular

File: api/models/Survey.js [hunk 22]
```diff
@@ -3,10 +3,18 @@ module.exports = {
...
+    },
+    job: {
+      model: 'job'
+    },
+    questions: {
+      collection: "question",
+      via: "survey"
     }
   },
   tableName: "surveys"
```
Comment (zaikin-andrew) [inline api/models/Survey.js]: singular

File: api/services/CandidateResponseService.js [hunk 23]
```diff
@@ -6,39 +6,33 @@ module.exports = {
...
+        if (!survey) return callback({status: 404, message: 'Survey not found'});
+        if (!survey.questions.length) return callback({status: 404, message: 'Not exist questions for this survey'});
+        cb(null, survey.questions);
       })
   }
 
-  function getOptionBuIdQiestion(questions, cb) {
-    Options.find({id_question: questions.filter((q) => q.type === "category").map((q) => q.id)})
-      .exec((err, results) => {
+  function getOptionBuIdQuestion(questions, cb) {
```
Comment (zaikin-andrew) [inline api/services/CandidateResponseService.js]: `By`

File: api/services/CandidateResponseService.js [hunk 24]
```diff
@@ -6,39 +6,33 @@ module.exports = {
 
 function saveResponses(jobId, surveyId, callback) {
   const inputData = MockService.getResponses();
-  async.waterfall([findJob, findSurvey, getQuestionsBySurveyID, getOptionBuIdQiestion, workWithResponses], resultFind);
+  async.waterfall([findJob, getQuestionBySurveyId, getOptionBuIdQuestion], finalWork);
```
Comment (zaikin-andrew) [inline api/services/CandidateResponseService.js]: One function - one action. I will have to write unit tests in future. Keep it in mind.
You wrote functions for other models. Why are they here? Move to a suitable place and import them.

File: api/services/helper.js [hunk 25]
```diff
@@ -10,3 +11,11 @@ function sendResult(res){
     if(!result) return res.ok();
   }
 }
+
+function parseJson(data) {
+  try {
+    return JSON.parse(data);
+  } catch (e) {
```
Comment (zaikin-andrew) [inline api/services/helper.js]: what about error handler?
What if you will have a huge JSON? This operation synchronous. You will get stuck node.

**Pull request merged**